### ISSUE TITLE
Disabled the Composer no-dev flag when dev dependencies are included

### ISSUE THREAD

**Issue by nunomaduro** [humbug/box#382]
## Bug report

| Question    | Answer
| ------------| ---------------
| Box version | 3.6.0
| PHP version | 7.3
| Platform with version | MacOS
| Github Repo | https://github.com/nunomaduro/exclude-dev-files-bug

Issue: The composer dev dependencies are not being included.

- The steps to reproduce your issue

```
git clone https://github.com/nunomaduro/exclude-dev-files-bug
cd exclude-dev-files-bug
composer update
cd builds
php application
```

- The exact command run for compilation:
```./box compile --working-dir="/Users/nunomaduro/Temp/exclude-dev-files-bug" --config="/Users/nunomaduro/Temp/exclude-dev-files-bug/box.json"```
- Your configuration file
```json
{
    "chmod": "0755",
    "directories": [
        "app",
        "bootstrap",
        "config",
        "vendor"
    ],
    "files": [
        "composer.json"
    ],
    "exclude-dev-files": false,
    "exclude-composer-files": false,
    "compression": "GZ",
    "compactors": [
        "KevinGH\\Box\\Compactor\\Php",
        "KevinGH\\Box\\Compactor\\Json"
    ]
}
```

- The exact command run for execution:
```
php builds/application
```

Result:
```
PHP Fatal error:  Uncaught Error: Class 'LaravelZero\Framework\Application' not found in phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/bootstrap/app.php:14
Stack trace:
#0 phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/application(19): require_once()
#1 /Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application(14): require('phar:///Users/n...')
#2 {main}
  thrown in phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/bootstrap/app.php on line 14

Fatal error: Uncaught Error: Class 'LaravelZero\Framework\Application' not found in phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/bootstrap/app.php:14
Stack trace:
#0 phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/application(19): require_once()
#1 /Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application(14): require('phar:///Users/n...')
#2 {main}
  thrown in phar:///Users/nunomaduro/Temp/exclude-dev-files-bug/builds/application/bootstrap/app.php on line 14

```

**Comment by theofidry** [comment]
Thanks for the report. A few notes:

- You shouldn't give the `config="/Users/nunomaduro/Temp/exclude-dev-files-bug/box.json"` argument IMO: Box will take in order `box.json` and then `box.json.dist`. So providing a `box.json.dist` file gives more flexibility to the user to for example have a `box.json` file without compression for debug purposes.
- `"composer.json"` is always "collected" when found. But if you want to include the composer files in the PHAR [`exclude-composer-files: false`](https://github.com/humbug/box/blob/master/doc/configuration.md#excluding-the-composer-files-exclude-composer-files) is more appropriate

Back on the original issue: turns out the problem is dumping the Composer autoloader which has the `no-dev` flag. I'll try to do a PR soon to disable that flag when `exclude-dev-files` is set to false as otherwise it will never work if the composer autoloader is dumped.

**Comment by theofidry** [comment]
I'm trying to have a deeper look as to me there is something more as I don't see why `LaravelZero\Framework\Application` isn't included with the `no-dev` flag

**Comment by theofidry** [comment]
Hm after having another look, could you tell me why `laravel-zero/framework` is a dev dependency? It doesn't make any sense to me

**Comment by nunomaduro** [comment]
Because after compiling the whole application into a single file, I want to deploy a 0 dependencies package.

**Comment by nunomaduro** [comment]
I can already do it, but I need to:

1. Move dependencies to required,
2. Compile the PHAR.
3. Move again dependencies to dev-required.

Here is an example here I have done this: https://github.com/laravel-zero/installer.

**Comment by theofidry** [comment]
>Because after compiling the whole application into a single file, I want to deploy a 0 dependencies package.

I don't think it's a good idea. If your application depends on `acme/foo`, then this dependency needs to be a regular (as non-dev) dependency. Otherwise the Composer autoloader is completely broken. You will be able to install your package just fine without adding more dependencies, but the installed code will not work.

>Because after compiling the whole application into a single file, I want to deploy a 0 dependencies package.

There is 2 ways to do this:

- Provide a PHAR or an equivalent
- Provide a customized Composer package, for which a non-exhaustive list of solutions is:
  - The package just provides the PHAR binary, i.e. it's a tiny package just to distribute the PHAR as a regular Composer package
  - Have your package code with the vendor all declared as source code: your code will effectively have no dependency anymore although you are way more subject to code conflicts
  - Like the point above but leveraging PHP-Scoper which will remove the potential code conflicts

**Comment by nunomaduro** [comment]
Thank you for this @theofidry.

**Comment by mipon** [comment]
Hi there,

@theofidry Is this already solved? 

I've got into a situation that Composer `require-dev` items are not loaded properly. Files are there in the compiled phar file.  However, `vendor/composer/autoload-classmap.php` are different between the composer one and the compiled phar one. 

I had to spend a whole day trying to figure this out. If I move those `require-dev` items to `require`, everything goes fine. 

```bash
$ php box.phar --version
Box version 3.8.4@120b0a3 2019-12-13 17:22:43 UT
```
